src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs

```diff
@@ -0,0 +1,700 @@
+            var desList = new List<IAsyncDisposable>();
+            try
+            {
+                using (var initializeResult = await _source.InitializeAsync(_options, result))
+                {
+                    foreach (var des in _destinations)
+                    {
+                        desList.Add(des.InitializeAsync(_options, result, initializeResult.BuildNumber));
+                    }
+
+                    ModifyV41OperateOnTypes(initializeResult.BuildNumber, isLastFile);
+
+                    var buildType = BuildVersion.Type(initializeResult.BuildNumber);
+                    var currentType = await _source.GetNextTypeAsync();
+                    while (currentType != DatabaseItemType.None)
+                    {
+                        await ProcessTypeAsync(currentType, result, buildType, ensureStepsProcessed);
+
+                        currentType = await _source.GetNextTypeAsync();
+                    }
+
+                    if (ensureStepsProcessed)
+                    {
+                        EnsureProcessed(result);
+                    }
+
+                    return result;
+                }
+            }
+            finally
+            {
+                foreach (var des in desList)
+                {
+                    await des.DisposeAsync();
+                }
+            }
+        }
+        
+        protected override async Task<SmugglerProgressBase.DatabaseRecordProgress> ProcessDatabaseRecordAsync(SmugglerResult result)
+        {
+            var action = new DatabaseRecordActions(_server, _databaseRecord, _databaseRecord.DatabaseName, LoggingSource.Instance.GetLogger<DatabaseDestination>(_databaseRecord.DatabaseName));
+
+            var databaseRecord = await _source.GetDatabaseRecordAsync();
+            _token.ThrowIfCancellationRequested();
+            //TODO - OnDatabaseRecordAction
+            result.DatabaseRecord.ReadCount++;
+            try
+            {
+                await action.WriteDatabaseRecordAsync(databaseRecord, result.DatabaseRecord, _options.AuthorizationStatus, _options.OperateOnDatabaseRecordTypes);
+            }
+            catch (Exception e)
+            {
+                result.AddError($"Could not write database record: {e.Message}");
+                result.DatabaseRecord.ErroredCount++;
+                throw;
+            }
+            return result.DatabaseRecord;
+        }
+
+        protected override async Task<SmugglerProgressBase.Counts> ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType)
+        {
+            var actionsList = new List<IDocumentActions>();
+            var throwOnCollectionMismatchError = _options.OperateOnTypes.HasFlag(DatabaseItemType.Tombstones) == false;
+            try
+            {
+                foreach (var des in _destinations)
+                {
+                    actionsList.Add(des.Documents(throwOnCollectionMismatchError));
```

Comment:
note that we need to ensure that if we have a big batch of documents for a particular shard, we aren't timing out the request to the others.
Can keep sending `' '`  or something like that, I think